### ISSUE TITLE
VK_EXT_debug_utils nor VK_EXT_debug_marker need a way to query whether frame is actively captured.

### ISSUE THREAD

**Issue by alecazam** [baldurk/renderdoc#2205]
## Description

When we inject groups and markers around source code all the time, we see the slowdown from the capture layer on sources.  On one platform, I see as much as 40ms from leaving markers on with the RenderDoc capture even though we're just running and not capturing the build.  

I can see some resource overhead from keeping ref counts alive on resources in case there's a capture, but I wasn't expecting this much of a slowdown coming from PIX and other capture tools.  I believe we only have access to extensions on some platforms, and can't integrate RenderDoc API directly.  So this would need to be supplied in an extension.

On Metal, there is an isCapturing call that we can tie all of our code to.  Something similar needs to be exposed in these extensions.

```
// called once per frame to see if capture is enabled
void Foo::BeginFrame()
{
	MTLCaptureManager* captureManager = [MTLCaptureManager sharedCaptureManager];
	m_capturingGPUFrame = captureManager.isCapturing;
}

// this is tested before calling our wrappers to fire off
// pVkCmdBeginDebugUtilsLabelEXT and pVkCmdEndDebugUtilsLabelEXT
bool Foo::IsDebugMarkerEnabled() const
{
	return g_renderer->m_capturingGPUFrame;
}
```

## Steps to reproduce

Do a capture with a lot of groups and markers in it.
Note perf with and without perf markers.  
Disable perf layer, and not 40ms faster.   This is platform specific.

## Environment

* RenderDoc version: v1.12
* Operating System: Windows 10 Pro
* Graphics API: Vulkan

**Comment by baldurk** [comment]
I would be _very_ surprised if the overhead of command buffer markers was measurable, certainly not to the degree of 40ms. Not unless you're doing an obscene number of those calls. There's not much that can be done to optimise those calls and I don't want to start trying to micro-optimising unless I'm sure that's the problem.

Can you please share a reproducible test case where there's high overhead from the marker calls so that I can investigate in detail?

**Comment by alecazam** [comment]
I think the capture slowdown is actually the accumulation of commands on this platform.  I can't share the code or platform in this case.  It's more complex when capture is always on, and you want testers to be able to send RenderDoc captures easily.  

But I also think extensions should offer ways where you can optimize out function and string work during non-capture situations.  Anyways, I liked that Metal offered a test for this, and we use that call to stop injecting any strings into the command buffer until a capture occurs.

In reading about RenderDoc, there was only a test for this that worked on Windows only and not on any other platforms.  If this were formalized into the Vulkan extension specs, then all systems could just call that.

**Comment by baldurk** [comment]
I'm afraid I don't understand what you're talking about, what accumulation of commands?

If this only happens on one platform then it doesn't sound like a RenderDoc bug to me, since the same code runs on all platforms and will have effectively the same overhead. Maybe there's some other layer that's slowing down unnecessarily on these commands.

I'm also not sure what test you're talking about, there is no windows-only test. With a handful of exceptions RenderDoc features work on any platform where possible, certainly anything relevant to debug markers or API interfaces.

However calling the marker label functions unconditionally (except in e.g. a release build) should have negligible performance overhead. It doesn't sound like you've measured what the problem is, trying to avoid calling these functions except mid frame capture is a hack because these functions shouldn't be expensive in the first place. It also risks missing some command buffers if they start recording before the capture begins.

**Comment by alecazam** [comment]
No, I isolated it to enabling the RenderDoc capture extension on this platform and then not setting that.  We ended up only enabling this extension at startup off a flag when capturing.  I was hoping to leave the capture on for testing, but my colleagues weren't impressed by the CPU hit from our perf tools.  I haven't compared perf on other platforms to see if the CPU overhead is there yet.  

I saw this comment on the "In Application API" but I guess it's just about dynamic loading of RenderDoc.  "This is not supported on Linux or Android due to the differing methods of hooking and only Windows supporting injection at runtime".

It looks like the RenderDoc api already has this, but I don't think we have access to this API.
uint32_t IsFrameCapturing()

I didn't say the markers were the cause of the slowdown above, but I just liked the low call overhead from perf tools like Metal GPU capture when capture is not active.   I can only assume the slowdown we see is the capture of command and resource on every frame by the capture itself.  Even though we only snap a one or less frame in a run.  Sorry, but we don't have the RenderDoc sources in this case on this platform, so I'm guessing at the CPU slowdown.

**Comment by baldurk** [comment]
OK it's very misleading to say that you see slowdown from leaving markers enabled when actually it sounds like markers have nothing to do with it and you're seeing general overhead from launching your program from RenderDoc. I don't expect it would make any difference at all if you were able to enable or disable markers only when capturing because the overhead will be coming largely from other places, and honestly I would expect that to be true for all tools. Depending on the platform and how the tool works it may have different overheads while capturing or idle, it's hard to compare without specific examples.

I do profile and optimise whenever possible to keep the overhead to a minimum where possible and improve things, but I do need to be able to run an actual or representative workload to be able to see where the problems are. In the past few versions I've applied optimisations based on significantly complex vulkan applications and I've been able to limit the background overhead to a reasonable degree. Generally I aim for single-digit millisecond overhead but there will always be some and it depends on what the application does, it's quite possible to hit nasty API cases that need to be specifically optimised for. Some of these optimisations are application and GPU specific, so it would help to know which platform are you talking about. Is it an android device of some kind?

The in-application API is indeed available on every platform. The comment you found is specifically about late-loading the renderdoc module which is only supported on windows, but every platform can query for the appropriate renderdoc module and access the API. However as I say above I believe you are completely on the wrong track at trying to disable the markers because I don't expect it will make even a fraction of a difference.
